httpql-core/src/main/java/com/hubspot/httpql/core/ann/FilterJoinByDescriptor.java

```diff
@@ -0,0 +1,19 @@
+package com.hubspot.httpql.core.ann;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({
+    ElementType.FIELD, ElementType.METHOD
+})
+public @interface FilterJoinByDescriptor {
+  /**
+   * The full class name of the filter implementation like com.hubspot.lib.data.MyFilterDescriptor
+   */
+  String value();
```

Comment:
We could use service loading for this too like with `@Filter`
`Class<? extends JoinDescriptor> value();`

With 
```java
public interface JoinDescriptor {}
```

```java
public interface JoinDescriptorImpl extends JoinDescriptor {

  Field<?> getField(FilterEntry filterEntry);

  JoinCondition getJoinCondition(QuerySpec querySpec);

  Class<? extends JoinDescriptor> getAnnotationClass();
}
```
...
`@FilterJoinByDescriptor(TagIdDescriptor.class)`

```java
interface TagIdDescriptor extends JoinDescriptor {}
```

```java
@AutoService(JoinDescriptorImpl.class)
class TagIdDescriptorImpl extends JoinDescriptorImpl {
  ...
  @Override
  public Class<? extends JoinDescriptor> getAnnotationClass() {
    return TagIdDescriptor.class;
  }
}
```

Reply:
That's neat. I wonder if it will work when the implementations are outside httpql. There are bunch of custom join descriptors in content4j and other repos.

Reply:
Yeah, it would work in a similar fashion. So long as the impl is on the classpath (the same way that it would need to be if doing it through `ClassLoader#loadClass(String)`).
You'd get better compile-time guarantees than when using a string. You still wouldn't be able to guarantee that the `FooDescriptorImpl` exists for `FooDescriptor` at compile-time